File: orchestra/runtime.py
"""Stock CLI runtimes for Claude Code and Codex.

The engine depends only on ``AgentRuntime``. Each implementation owns its own
session namespace because the two CLIs establish sessions differently: Claude
accepts a caller-provided id, while Codex emits its thread id after starting.
"""

from __future__ import annotations

import json
import os
import shutil
import signal
import subprocess
import sys
import time
from dataclasses import dataclass
from pathlib import Path

from .models import ERROR, OK, WARNING, HealthReport, RunSnapshot, StartResult
from .state import FileStateStore, lock_is_held, write_json_atomic

BASE_ENV_NAMES = (
    "HOME",
    "USER",
    "LOGNAME",
    "PATH",
    "SHELL",
    "TMPDIR",
    "LANG",
    "LC_ALL",
    "SSH_AUTH_SOCK",
)


@dataclass(frozen=True, slots=True)
class ClaudeRuntimeConfig:
    executable: str = "claude"
    model: str = ""
    permission_mode: str = "default"
    passthrough_env: tuple[str, ...] = ()
    base_env_names: tuple[str, ...] = BASE_ENV_NAMES
    extra_args: tuple[str, ...] = ()
    system_prompt_file: str = "SYSTEM.md"


@dataclass(frozen=True, slots=True)
class CodexRuntimeConfig:
    executable: str = "codex"
    model: str = ""
    sandbox: str = "workspace-write"
    network_access: bool = True
    approval_policy: str = "never"
    dangerously_bypass_approvals_and_sandbox: bool = False
    skip_git_repo_check: bool = True
    passthrough_env: tuple[str, ...] = ()
    base_env_names: tuple[str, ...] = BASE_ENV_NAMES
    extra_args: tuple[str, ...] = ()
    system_prompt_file: str = "SYSTEM.md"

    def __post_init__(self) -> None:
        if self.sandbox not in {"read-only", "workspace-write", "danger-full-access"}:
            raise ValueError(f"invalid Codex sandbox: {self.sandbox}")
        if self.approval_policy not in {
            "",
            "never",
            "on-request",
            "untrusted",
            "unless-trusted",
        }:
            raise ValueError(
                f"unsupported Codex approval policy: {self.approval_policy}"
            )


class _DetachedCliRuntime:
    name = "cli"

    def __init__(self, store: FileStateStore):
        self.store = store
        self._health = HealthReport(OK, "ready")

    def health(self) -> HealthReport:
        return self._health

    def is_alive(self, task_dir: Path) -> bool:
        return lock_is_held(self.store.state_path(task_dir, "turn.lock"))

    def _spawn(
        self,
        task_dir: Path,
        prompt: str,
        resume: bool,
        sequence: int,
        *,
        executable_name: str,
        worker_name: str,
        base_env_names: tuple[str, ...],
        passthrough_env: tuple[str, ...],
        worker_environment: dict[str, str],
    ) -> StartResult:
        executable = shutil.which(executable_name) or executable_name
        if not Path(executable).is_file() and shutil.which(executable) is None:
            detail = f"agent executable not found: {executable_name}"
            self._health = HealthReport(ERROR, detail)
            return StartResult(False, detail)

        state_dir = self.store.state_path(task_dir)
        write_json_atomic(
            state_dir / "turn.json",
            {
                "version": 1,
                "sequence": sequence,
                "pid": 0,
                "started_at": time.time(),
                "resumed": resume,
                "prompt": prompt,
                "runtime": self.name,
            },
        )
        (state_dir / "status.json").unlink(missing_ok=True)
        environment = {
            name: os.environ[name]
            for name in base_env_names + passthrough_env
            if name in os.environ
        }
        environment.update(worker_environment)
        environment["ORCHESTRA_AGENT_BIN"] = executable
        environment["ORCHESTRA_PROMPT"] = prompt
        environment["ORCHESTRA_RESUME"] = "1" if resume else "0"
        environment["ORCHESTRA_TURN_SEQUENCE"] = str(sequence)
        worker = Path(__file__).with_name(worker_name)
        try:
            with (state_dir / "spawn.log").open("ab") as log:
                process = subprocess.Popen(
                    [sys.executable, str(worker)],
                    cwd=str(task_dir),
                    env=environment,
                    stdout=log,
                    stderr=log,
                    start_new_session=True,
                )
        except OSError as error:
            detail = f"spawn failed: {error}"
            self._health = HealthReport(WARNING, detail)
            return StartResult(False, detail)
        turn = self.store.turn(task_dir)
        turn["pid"] = process.pid
        write_json_atomic(state_dir / "turn.json", turn)
        self._health = HealthReport(OK, f"started task turn {sequence}")
        return StartResult(True)

    def stop(self, snapshot: RunSnapshot, *, force: bool = False) -> None:
        pid = int(snapshot.turn.get("pid") or 0)
        if not pid:
            return
        try:
            os.killpg(pid, signal.SIGKILL if force else signal.SIGTERM)
        except ProcessLookupError:
            return
        except PermissionError as error:
            self._health = HealthReport(WARNING, f"could not stop pid {pid}: {error}")


class ClaudeCodeRuntime(_DetachedCliRuntime):
    name = "claude-code"

    def __init__(self, store: FileStateStore, config: ClaudeRuntimeConfig):
        super().__init__(store)
        self.config = config

    def start(
        self,
        task_dir: Path,
        prompt: str,
        resume: bool,
        sequence: int,
    ) -> StartResult:
        session_id = self.store.session_id(task_dir, self.name)
        resume = resume or self._has_transcript(task_dir, session_id)
        return self._spawn(
            task_dir,
            prompt,
            resume,
            sequence,
            executable_name=self.config.executable,
            worker_name="_worker.py",
            base_env_names=self.config.base_env_names,
            passthrough_env=self.config.passthrough_env,
            worker_environment={
                "ORCHESTRA_SESSION_ID": session_id,
                "ORCHESTRA_MODEL": self.config.model,
                "ORCHESTRA_PERMISSION_MODE": self.config.permission_mode,
                "ORCHESTRA_EXTRA_ARGS": json.dumps(self.config.extra_args),
                "ORCHESTRA_SYSTEM_PROMPT_FILE": self.config.system_prompt_file,
            },
        )

    @staticmethod
    def _has_transcript(task_dir: Path, session_id: str) -> bool:
        mangled = str(Path(task_dir).resolve()).replace("/", "-")
        return (
            Path.home()
            / ".claude"
            / "projects"
            / mangled
            / f"{session_id}.jsonl"
        ).is_file()


class CodexCliRuntime(_DetachedCliRuntime):
    name = "codex-cli"

    def __init__(self, store: FileStateStore, config: CodexRuntimeConfig):
        super().__init__(store)
        self.config = config

    def start(
        self,
        task_dir: Path,
        prompt: str,
        resume: bool,
        sequence: int,
    ) -> StartResult:
        session_id = self.store.session_id(task_dir, self.name, create=False)
        resume = bool(
            session_id
            and self.store.session_ready(task_dir, self.name, session_id)
        )
        session_path = self.store.session_path(task_dir, self.name)
        ready_path = self.store.session_ready_path(task_dir, self.name)
        return self._spawn(
            task_dir,
            prompt,
            resume,
            sequence,
            executable_name=self.config.executable,
            worker_name="_codex_worker.py",
            base_env_names=self.config.base_env_names,
            passthrough_env=self.config.passthrough_env,
            worker_environment={
                "ORCHESTRA_SESSION_ID": session_id,
                "ORCHESTRA_SESSION_PATH": str(session_path),
                "ORCHESTRA_SESSION_READY_PATH": str(ready_path),
                "ORCHESTRA_MODEL": self.config.model,
                "ORCHESTRA_CODEX_SANDBOX": self.config.sandbox,
                "ORCHESTRA_CODEX_NETWORK_ACCESS": (
                    "1" if self.config.network_access else "0"
                ),
                "ORCHESTRA_CODEX_APPROVAL_POLICY": self.config.approval_policy,
                "ORCHESTRA_CODEX_DANGEROUS_BYPASS": (
                    "1"
                    if self.config.dangerously_bypass_approvals_and_sandbox
                    else "0"
                ),
                "ORCHESTRA_CODEX_SKIP_GIT_CHECK": (
                    "1" if self.config.skip_git_repo_check else "0"
                ),
                "ORCHESTRA_EXTRA_ARGS": json.dumps(self.config.extra_args),
                "ORCHESTRA_SYSTEM_PROMPT_FILE": self.config.system_prompt_file,
            },
        )

File: patchwake/runtime.ts
/** Detached CLI turns. The child inherits the OS lock before intent is published. */
import {
  accessSync,
  closeSync,
  constants,
  existsSync,
  openSync,
  statSync,
  rmSync,
} from "node:fs";
import { spawn } from "node:child_process";
import { delimiter, isAbsolute, join, resolve } from "node:path";
import { homedir } from "node:os";
import { fileURLToPath } from "node:url";
import { health, OK, ERROR, WARNING } from "./models.js";
import type {
  HealthReport,
  RunSnapshot,
  StartResult,
  TurnRecord,
} from "./models.js";
import type { AgentRuntime } from "./ports.js";
import {
  FileStateStore,
  LockBusy,
  lockIsHeld,
  nowSeconds,
  openLock,
  writeJsonAtomic,
} from "./state.js";
import { errorMessage, hasCode, isRecord } from "./json.js";

export const BASE_ENV_NAMES = [
  "HOME",
  "USER",
  "LOGNAME",
  "PATH",
  "SHELL",
  "TMPDIR",
  "LANG",
  "LC_ALL",
  "SSH_AUTH_SOCK",
] as const;
interface CommonRuntimeConfig {
  readonly executable?: string;
  readonly model?: string;
  readonly passthroughEnv?: readonly string[];
  readonly baseEnvNames?: readonly string[];
  readonly extraArgs?: readonly string[];
  readonly systemPromptFile?: string;
}
export interface ClaudeRuntimeConfig extends CommonRuntimeConfig {
  readonly permissionMode?: string;
}
export interface CodexRuntimeConfig extends CommonRuntimeConfig {
  readonly sandbox?: "read-only" | "workspace-write" | "danger-full-access";
  readonly networkAccess?: boolean;
  readonly approvalPolicy?:
    "" | "never" | "on-request" | "untrusted" | "unless-trusted";
  readonly dangerouslyBypassApprovalsAndSandbox?: boolean;
  readonly skipGitRepoCheck?: boolean;
}
function findExecutable(name: string): string | null {
  const paths =
    isAbsolute(name) || name.includes("/")
      ? [resolve(name)]
      : (process.env.PATH ?? "")
          .split(delimiter)
          .map((directory) => resolve(directory, name));
  for (const path of paths) {
    try {
      accessSync(path, constants.X_OK);
      if (statSync(path).isFile()) return path;
    } catch (error) {
      if (!hasCode(error, "ENOENT", "EACCES", "ENOTDIR")) throw error;
    }
  }
  return null;
}

abstract class DetachedCliRuntime implements AgentRuntime {
  abstract readonly name: string;
  protected current = health(OK, "ready");
  constructor(readonly store: FileStateStore) {}
  health(): HealthReport {
    return this.current;
  }
  isAlive(taskDir: string): boolean {
    return lockIsHeld(this.store.statePath(taskDir, "turn.lock"));
  }
  abstract start(
    taskDir: string,
    prompt: string,
    resume: boolean,
    sequence: number,
  ): Promise<StartResult>;

  protected async spawnTurn(
    taskDir: string,
    prompt: string,
    resume: boolean,
    sequence: number,
    config: CommonRuntimeConfig,
    executableName: string,
    workerEnvironment: Record<string, string>,
  ): Promise<StartResult> {
    const executable = findExecutable(executableName);
    if (!executable) {
      const detail = `agent executable not found: ${executableName}`;
      this.current = health(ERROR, detail);
      return { ok: false, detail };
    }
    let lockFd: number;
    try {
      lockFd = openLock(this.store.statePath(taskDir, "turn.lock"));
    } catch (error) {
      if (error instanceof LockBusy)
        return { ok: false, detail: "another turn holds the lock" };
      throw error;
    }
    try {
      const intent: TurnRecord = {
        version: 1,
        sequence,
        pid: 0,
        started_at: nowSeconds(),
        resumed: resume,
        prompt,
        runtime: this.name,
      };
      writeJsonAtomic(this.store.statePath(taskDir, "turn.json"), intent);
      rmSync(this.store.statePath(taskDir, "status.json"), { force: true });
      const env: NodeJS.ProcessEnv = {};
      for (const name of [
        ...(config.baseEnvNames ?? BASE_ENV_NAMES),
        ...(config.passthroughEnv ?? []),
      ]) {
        if (process.env[name] !== undefined) env[name] = process.env[name];
      }
      Object.assign(env, workerEnvironment, {
        PATCHWAKE_RUNTIME: this.name,
        PATCHWAKE_AGENT_BIN: executable,
        PATCHWAKE_PROMPT: prompt,
        PATCHWAKE_RESUME: resume ? "1" : "0",
        PATCHWAKE_TURN_SEQUENCE: String(sequence),
        PATCHWAKE_MODEL: config.model ?? "",
        PATCHWAKE_EXTRA_ARGS: JSON.stringify(config.extraArgs ?? []),
        PATCHWAKE_SYSTEM_PROMPT_FILE: config.systemPromptFile ?? "SYSTEM.md",
      });
      const logFd = openSync(this.store.statePath(taskDir, "spawn.log"), "a");
      let child;
      try {
        child = spawn(
          process.execPath,
          [fileURLToPath(new URL("./worker.js", import.meta.url))],
          {
            cwd: taskDir,
            env,
            detached: true,
            stdio: ["ignore", logFd, logFd, "ipc", lockFd],
          },
        );
      } catch (error) {
        const detail = `spawn failed: ${errorMessage(error)}`;
        this.current = health(WARNING, detail);
        return { ok: false, detail };
      } finally {
        closeSync(logFd);
      }
      const result = await new Promise<StartResult>((resolveResult) => {
        let settled = false;
        const finish = (value: StartResult): void => {
          if (settled) return;
          settled = true;
          clearTimeout(timer);
          if (child.connected) child.disconnect();
          child.unref();
          resolveResult(value);
        };
        const timer = setTimeout(() => {
          if (child.pid) {
            try {
              process.kill(-child.pid, "SIGKILL");
            } catch (error) {
              if (!hasCode(error, "ESRCH"))
                this.current = health(WARNING, errorMessage(error));
            }
          }
          finish({
            ok: false,
            detail: "worker did not acknowledge agent startup",
          });
        }, 10_000);
        child.once("error", (error) =>
          finish({ ok: false, detail: `spawn failed: ${error.message}` }),
        );
        child.once("exit", (code, signal) =>
          finish({
            ok: false,
            detail: `worker exited before startup: ${code ?? signal}`,
          }),
        );
        child.once("message", (message: unknown) => {
          if (isRecord(message) && message.type === "started")
            finish({ ok: true });
          else
            finish({
              ok: false,
              detail:
                isRecord(message) && typeof message.detail === "string"
                  ? message.detail
                  : "invalid worker acknowledgement",
            });
        });
        // Register errors before writing the PID, since spawn failure is asynchronous.
        try {
          writeJsonAtomic(this.store.statePath(taskDir, "turn.json"), {
            ...intent,
            pid: child.pid ?? 0,
          });
        } catch (error) {
          if (child.pid) {
            try {
              process.kill(-child.pid, "SIGKILL");
            } catch {
              /* It may already have exited. */
            }
          }
          finish({
            ok: false,
            detail: `could not persist worker PID: ${errorMessage(error)}`,
          });
        }
      });
      this.current = result.ok
        ? health(OK, `started task turn ${sequence}`)
        : health(WARNING, result.detail ?? "spawn failed");
      return result;
    } finally {
      closeSync(lockFd);
    }
  }
  stop(
    snapshot: RunSnapshot,
    { force = false }: { force?: boolean } = {},
  ): void {
    const pid = snapshot.turn.pid ?? 0;
    // A stale PID without a held lock must never be signalled.
    if (pid <= 0 || !this.isAlive(snapshot.directory)) return;
    try {
      process.kill(-pid, force ? "SIGKILL" : "SIGTERM");
    } catch (error) {
      if (hasCode(error, "ESRCH")) return;
      if (hasCode(error, "EPERM", "EACCES")) {
        this.current = health(
          WARNING,
          `could not stop pid ${pid}: ${errorMessage(error)}`,
        );
        return;
      }
      throw error;
    }
  }
}
export class ClaudeCodeRuntime extends DetachedCliRuntime {
  readonly name = "claude-code";
  constructor(
    store: FileStateStore,
    readonly config: ClaudeRuntimeConfig = {},
  ) {
    super(store);
  }
  start(
    taskDir: string,
    prompt: string,
    resume: boolean,
    sequence: number,
  ): Promise<StartResult> {
    const sessionId = this.store.sessionId(taskDir, this.name);
    const transcript = join(
      homedir(),
      ".claude",
      "projects",
      resolve(taskDir).replaceAll("/", "-"),
      `${sessionId}.jsonl`,
    );
    return this.spawnTurn(
      taskDir,
      prompt,
      resume || existsSync(transcript),
      sequence,
      this.config,
      this.config.executable ?? "claude",
      {
        PATCHWAKE_SESSION_ID: sessionId,
        PATCHWAKE_PERMISSION_MODE: this.config.permissionMode ?? "default",
      },
    );
  }
}
export class CodexCliRuntime extends DetachedCliRuntime {
  readonly name = "codex-cli";
  constructor(
    store: FileStateStore,
    readonly config: CodexRuntimeConfig = {},
  ) {
    super(store);
    if (
      !["read-only", "workspace-write", "danger-full-access"].includes(
        config.sandbox ?? "workspace-write",
      )
    )
      throw new Error(`invalid Codex sandbox: ${config.sandbox}`);
    if (
      !["", "never", "on-request", "untrusted", "unless-trusted"].includes(
        config.approvalPolicy ?? "never",
      )
    )
      throw new Error(
        `unsupported Codex approval policy: ${config.approvalPolicy}`,
      );
  }
  start(
    taskDir: string,
    prompt: string,
    _resume: boolean,
    sequence: number,
  ): Promise<StartResult> {
    const sessionId = this.store.sessionId(taskDir, this.name, false);
    const resume = this.store.sessionReady(taskDir, this.name, sessionId);
    return this.spawnTurn(
      taskDir,
      prompt,
      resume,
      sequence,
      this.config,
      this.config.executable ?? "codex",
      {
        PATCHWAKE_SESSION_ID: sessionId,
        PATCHWAKE_SESSION_PATH: this.store.sessionPath(taskDir, this.name),
        PATCHWAKE_SESSION_READY_PATH: this.store.sessionReadyPath(
          taskDir,
          this.name,
        ),
        PATCHWAKE_CODEX_SANDBOX: this.config.sandbox ?? "workspace-write",
        PATCHWAKE_CODEX_NETWORK_ACCESS:
          (this.config.networkAccess ?? true) ? "1" : "0",
        PATCHWAKE_CODEX_APPROVAL_POLICY: this.config.approvalPolicy ?? "never",
        PATCHWAKE_CODEX_DANGEROUS_BYPASS: this.config
          .dangerouslyBypassApprovalsAndSandbox
          ? "1"
          : "0",
        PATCHWAKE_CODEX_SKIP_GIT_CHECK:
          (this.config.skipGitRepoCheck ?? true) ? "1" : "0",
      },
    );
  }
}
